DEVDOCS-1341 (#652)
bigcommerce/dev-docs: docs/stencil-docs/page-builder/configuration.md

* Keys located in both configuration files'`settings` define the theme's look, feel, and functionality.
* Keys located in the `variations` object of `config.json` define variations of the theme. There is no limit to the number of variations each theme can include.
* For front-matter properties to be editable, your theme's Handlebars template must call certain Handlebars helpers to transform the `config.json` entries into JavaScript values.
* For fonts to be editable, a Sass stylesheet must call certain custom Sass functions to transform the `config.json` font entries into CSS values.
* For styles to be editable, a Sass stylesheet must call certain custom Handlebars helpers to transform the `config.json` entries into CSS values.

For documentation on the Cornerstone theme's principal keys, see Cornerstone's [config.json Metadata](https://github.com/bigcommerce/cornerstone/blob/master/config.json) and [schema.json Metadata](https://github.com/bigcommerce/cornerstone/blob/master/schema.json) entries. 

### File management requirements

> * For any theme setting, such as a Sass variable or a front-matter value, to be merchant-customizable, that setting and its values must be present in `schema.json`. You have to manually provide these declarations, according to the structure described here.

Reviewer: Clarify where the "...structure described here" is referring to.

> * Each key that you create in `schema.json` must have a corresponding `config.json` key whose name matches its ID value. This `config.json` key sets the default value even if that is an empty string. A `schema.json` setting without an `id`-matched `config.json` key will not appear to users in the Page Builder UI.

Reviewer: ```suggestion
Each key that you create in `schema.json` must have a corresponding `config.json` key whose name matches its ID value. This `config.json` key sets the default value even if that is an empty string. A `schema.json` setting without an `id`-matched `config.json` key will not appear to users in the Page Builder UI.
```

* Merchants must select one variation to edit at a time. The selections that they make in the remainder of Page Builder's UI will apply only to that selected variation.
* Page Builder's remaining sequence of top-level section headings corresponds directly to the top-level section objects that you declare in `schema.json`. The options displayed within these expandable section headings correspond directly to the key-value pairs that you nest within `schema.json`'s corresponding section objects.

In all, the structure that you give your theme's `config.json` and `schema.json` files directly governs the UI that Page Builder exposes to merchants. 

## Page Builder data types

You are free to decide which properties of your theme to make editable in Page Builder and in which order to display them. Page Builder can expose any set of properties as long as your `schema.json` declares them using the data types that Page Builder supports.

Page Builder supports the following data types:
* color
* font

> * select [drop-down list]

Reviewer: ```suggestion
* select (drop-down list)
```

```json
{
    "name": "Colors",
    "settings": [
      {
        "type": "heading",
        "content": "General"
      },
      {
        "type": "color",
        "label": "Text Color",
        "id": "body-font-color"
      },
    {...}
     ]
}
```

These `"type": "heading"` statements do not reference data types. Rather, they declare display captions for the Page Builder UI's subcategories – the middle level nested within the section headings, but outside the individual options from which merchants can select. Inner options are designated by `"label": <label-text>` statements.

## Page Builder data structure in `schema.json` 

The `schema.json` nesting structure maps directly to the Page Builder UI displayed to merchants. Below the `variations` section, which contains data imported from `config.json`, the order and nesting of Page Builder's UI options directly match the order and nesting of your `schema.json` entries.

## Best practices

To prevent errors upon theme upload and avoid possible loss of customizations made through the Page Builder UI at runtime, please follow these guidelines:

> * **Single-instance restriction per storefront.** We strongly recommend opening only one instance of Page Builder at a time against each storefront. We give this recommendation because there is currently no synchronization mechanism to reconcile configuration changes made by multiple Page Builder instances. `schema.json` will record the last changes made by any instance, but changes saved earlier by other instances might be lost.

Reviewer: ```suggestion
* **Single-instance restriction per storefront.** We strongly recommend opening only one instance of Page Builder at a time against each storefront. We recommend this because there is currently no synchronization mechanism to reconcile configuration changes made by multiple Page Builder instances. The `schema.json` file will record the last changes made by any instance, but changes saved earlier by other instances might be lost.
```

| ----------- | ----------- | 
| Adding a new key | Adding new keys is generally acceptable. However, each key in `schema.json` must have a matching default in `config.json`. |
 Deleting a key | Use caution when deleting a key. Doing so can break your new theme version's backward compatibility. | 
 Renaming a key | We do not recommend renaming keys. Instead, we recommend introducing a new key while maintaining the old key until there is no longer a need to use an older version of your theme.|
 Removing a variation| Each object within your `config.json` file's `variations` object defines one theme variation. If you adapt an existing theme and consciously want to remove one or more variations, you can remove the corresponding key(s). |

## Persistent settings storage

When store administrators use Page Builder to customize your store's theme, they save the store's resulting configuration settings to a separate configuration service at BigCommerce.

## Theme upgrades and settings

When a merchant upgrades your theme to a newer version, they carry forward all key-value pairs saved to the BigCommerce configuration service. For example, assume this customization/upgrade scenario:

> You release your Star Glow theme, version 1. This theme's `config.json` includes a key named `logo_size`, establishing a default value of 100x250. The combination of the key and the value composes a `logo_size` setting.

Reviewer: ```suggestion
You release your Star Glow theme, version 1.0. This theme's `config.json` includes a key named `logo_size`, establishing a default value of 100 x 250px. The combination of the key and the value composes a `logo_size` setting.

```

> The merchant uses Store Design to change the `logo_size` setting to 175x275 and stores this customized setting in the BigCommerce configuration service.

Reviewer: ```suggestion
The merchant uses Store Design to change the `logo_size` setting to 175 x 275px and stores this customized setting in the BigCommerce configuration service.
```

> You release Star Glow, version 1.1. In this theme revision, you have changed the `logo_size` to 300x300.

Reviewer: ```suggestion
You release Star Glow, version 1.1. In this theme revision, you have changed the `logo_size` to 300 x 300px.
```

> When the merchant applies Star Glow version 1.1 to their store, their custom `logo_size` setting of 175x275 remains in effect.

Reviewer: ```suggestion
When the merchant applies Star Glow version 1.1 to their store, their custom `logo_size` setting of 175 x 275px remains in effect.
```

> If the merchant creates a second store and applies Star Glow version 1.1 to it, that store will not have a custom `logo_size` setting and will default to the new theme version's 300x300 value.

Reviewer: ```suggestion
If the merchant creates a second store and applies Star Glow version 1.1 to it, that store will not have a custom `logo_size` setting and will default to the new theme version's 300 x 300px value.
```

## Troubleshooting

You may experience an issue when setting up the Page Builder UI. We recommend that you check the terminal window where you started Stencil CLI for any unexpected behavior that you encounter while developing your Stencil theme. In some cases, the terminal will provide a lengthy error message specifying where to look for problems. For less-detailed error messages, see the below list of potential issues and diagnostic suggestions.

### Empty drop-down list in Page Builder panel

* **Symptom:** A drop-down list's outline appears below its configured label. However, the list appears to be empty.
* **Likely Cause:** The `schema.json` file does not list a default value specified in the theme's `config.json` file.
* **Resolution:** Update `schema.json` to include the `config.json` value.

### Configured control missing from Page Builder panel

* **Symptom:** A control that you have configured within `schema.json` is completely absent from the Page Builder UI.
* **Likely Cause:** The specified data type is one of the following: text, text area, radio [button], or image. Page Builder does not currently support these data types.

> * **Resolution:** Display the user option using one of the supported data types: color, font, select [drop-down list], or checkbox.

Reviewer: ```suggestion
* **Resolution:** Display the user option using one of the supported data types: color, font, select (drop-down list), or checkbox.
```

### Theme changes not saved from Page Builder UI

* **Symptom:** The storefront does not reflect changes saved in a browser's Page Builder panel.
* **Likely Cause:** Check whether Page Builder can customize the same storefront.

> * **Resolution:** We strongly recommend opening only one instance of Page Builder, at a time, per storefront. BigCommerce currently provides no synchronization mechanism for configuration changes from multiple Page Builder instances. So the storefront’s `schema.json` will record the last changes made by any instance – but changes saved earlier by other instances might be lost.

Reviewer: ```suggestion
* **Resolution:** We strongly recommend opening only one instance of Page Builder at a time per storefront. BigCommerce currently provides no synchronization mechanism for configuration changes from multiple Page Builder instances. So the storefront’s `schema.json` will record the last changes made by any instance, but changes saved earlier by other instances might be lost.
```

 Renaming a key | We do not recommend renaming keys. Instead, we recommend introducing a new key, while maintaining the old key until it is no longer in use by anyone using an older version of your theme.|
 Removing a variation| Each object within your `config.json` file's `variations` object defines one theme variation. If you are adapting an existing theme and consciously want to remove one or more variations, you can do so by removing the corresponding key(s). |

## Persistent settings storage

When store administrators use Page Builder to customize your theme for their store, the store's resulting configuration settings are saved to a separate configuration service at BigCommerce.

## Theme upgrades and settings

When a merchant upgrades your theme to a newer version, all key-value pairs that were saved to the BigCommerce configuration service are carried forward. For example, assume this customization/upgrade scenario:

You release your Star Glow theme, version 1. This theme's `config.json` includes a key named `logo_size`, establishing a default value of 100x250. The combination of the key and the value compose a `logo_size` setting.

> The merchant uses Store Design to change the `logo_size` setting to 175x275. This customized setting is stored in the BigCommerce configuration service.

Reviewer 2: ```suggestion
The merchant uses Store Design to change the `logo_size` setting to 175x275. You can store this customized setting in the BigCommerce configuration service.
```

You release Star Glow, version 1.1. In this theme revision, you have changed the `logo_size` to 300x300.
When the merchant applies Star Glow version 1.1 to their store, their custom `logo_size` setting of 175x275 remains in effect.

> If you the merchant creates a second store and applies Star Glow version 1.1 to it, that store will not have a custom `logo_size` setting and will default to the new theme version's 300x300 value.

Reviewer 2: ```suggestion
If the merchant creates a second store and applies Star Glow version 1.1 to it, that store will not have a custom `logo_size` setting and will default to the new theme version's 300x300 value.
```

## Troubleshooting

> You may experience an issue when setting up the Page Builder UI. For any unexpected behavior that you encounter while developing your Stencil theme, we recommend that you check the terminal window where you started Stencil CLI. In some cases, the terminal will provide a verbose error message specifying where to look for problems. For less-detailed error messages, see the below list of potential issues and diagnostic suggestions.

Reviewer 2: ```suggestion
We recommend that you check the terminal window where you started Stencil CLI for any unexpected behavior that you encounter while developing your Stencil theme.  In some cases, the terminal will provide a lengthy error message specifying where to look for problems. For less-detailed error messages, see the below list of potential issues and diagnostic suggestions.
```

### Empty drop-down list in Page Builder panel

> * **Symptom:** A drop-down list’s outline appears below its configured label. However, the list appears to be empty.

Reviewer 2: ```suggestion
* **Symptom:** A drop-down list's outline appears below its configured label. However, the list appears to be empty.
```

> * **Likely Cause:** A default value specified in the theme’s config.json file is not enumerated in the schema.json file.

Reviewer 2: ```suggestion
* **Likely Cause:** The schema.json file does not list a default value specified in the theme's config.json file.
```

* **Resolution:** Update schema.json to include the config.json value.

### Configured control missing from Page Builder panel

* **Symptom:** A control that you have configured within schema.json is completely absent from the Page Builder UI.
* **Likely Cause:** The specified “type” is one of: text, text area, radio [button], or image (Page Builder does not currently support these data types).
* **Resolution:** Display the user option via one of the supported data types: color, font, select [drop-down list], or checkbox.

### Theme changes not saved from Page Builder UI

> * **Symptom:** Changes saved in a browser’s Page Builder panel are not reflected in the storefront.

Reviewer 2: ```suggestion
* **Symptom:** The storefront does not display changes saved in a browser’s Page Builder panel.
```

* **Likely Cause:** Check whether Page Builder can customize the same storefront.

> * **Resolution:** We strongly recommend opening only one instance of Page Builder, at a time, per storefront. BigCommerce currently provides no synchronization mechanism for configuration changes from multiple Page Builder instances. So the storefront’s schema.json will record the last changes made by any instance – but changes saved earlier by other instances might be lost.

Reviewer 2: ```suggestion
* **Resolution:** We strongly recommend opening only one instance of Page Builder, at a time, per storefront. BigCommerce currently provides no synchronization mechanism for configuration changes from multiple Page Builder instances. So the storefront's schema.json will record the last changes made by any instance – but changes saved earlier by other instances might be lost.
```
